llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1074,6 +1074,35 @@ void AsmPrinter::emitFunctionHeader() {
 
 static bool needFuncLabelsForEH(const MachineFunction &MF);
 
+/// Returns true if function begin label should be emitted due to a
+/// BlockAddress used as an initializer for a GlobalVariable.
+static bool needFuncLabelForBlockAddress(const MachineFunction &MF) {
+  SmallVector<const Value *, 4> Users;
+
+  for (const auto &MBB : MF) {
+    if (!MBB.isIRBlockAddressTaken())
+      continue;
+
+    const auto *BB = MBB.getAddressTakenIRBlock();
+    for (const auto &U : BB->users()) {
+      if (isa<BlockAddress>(U))
+        Users.push_back(U);
+    }
+  }
+
+  while (!Users.empty()) {
+    const Value *V = Users.pop_back_val();
+    if (isa<GlobalVariable>(V))
+      return true;
+    if (isa<Instruction>(V))
+      continue;
+    for (const auto &U : V->users())
+      Users.push_back(U);
+  }
+
+  return false;
+}
```

Comment:
```suggestion
static bool needFuncLabelForBlockAddress(const MachineFunction &MF) {
  for (const auto &MBB : MF) {
    if (!MBB.isIRBlockAddressTaken())
      continue;

    for (const auto &U : MBB.getAddressTakenIRBlock()->users()) {
      if (isa<BlockAddress>(U))
        return true;
    }
  }
  return false;
}
```
Something like this instead of the more complex logic?

Reply:
That'll return true for something like
```
void foo(void) {
  void *p = &&bar;

  ...

bar:
  ...
}
```
which doesn't need the special function-type label as it's just a PCC-relative calculation at function run time.

Reply:
Agreed that can adds unnecessary labels, just wasn't sure if we need the extra checks. Happy to keep the current version.